tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,334 @@
+      element: Dict[Text, Dict[Text, tf.Tensor]],
+      spec: execution_spec.ExecutionSpec,
+      graph_name_extended: Text
+  ) -> Generator[Dict[Text, Dict[Text, tf.Tensor]], None, None]:
+    """Executes one subgraph layer.
+
+    To execute a subgraph layer, we need to prepare for a feed_dict by
+    extracting tensors from element. Then, we run the subgraph and store the
+    outputs to element.
+
+    Since we import `GraphDef` protos, all the nodes now associate with a
+    prefix "import/". Also, TensorFlow feed_dict and outputs accept tensor
+    names instead of node names. Hence, a conversion from node_name to
+    "import/node_name:0" is necessary (assumption #2: one output per node).
+
+    Args:
+      element: A dictionary from graph name extendeds to a dictionary from
+               tensor names to tensors. It stores the graph inputs and all the
+               previous specs' outputs for the current graph.
+      spec: An ExecutionSpec for a subgraph layer.
+      graph_name_extended: As discussed in ExecuteOneGraph, it is the extended
+                           name of the current graph.
+
+    Returns:
+      A dictionary from graph name extended to a dictionary from tensor names
+      to tensors. The outputs of this subgraph layer are added.
+    """
+    element = copy.deepcopy(element)
+    input_tensor_names = [_import_tensor_name(node_name)
+                          for node_name in spec.input_names]
+    output_tensor_names = [_import_tensor_name(node_name)
+                           for node_name in spec.output_names]
+    feed_dict = {tensor_name: element[graph_name_extended][tensor_name]
+                 for tensor_name in input_tensor_names}
+
+    outputs = []
+    with tf.compat.v1.Session(graph=tf.Graph()) as sess:
+      tf.import_graph_def(spec.subgraph)
+      outputs = sess.run(output_tensor_names, feed_dict=feed_dict)
+
+    for output_tensor_name, output_tensor in zip(output_tensor_names, outputs):
+      element[graph_name_extended][output_tensor_name] = output_tensor
+
+    yield element
+
+
+def _import_tensor_name(node_name: Text) -> Text:
+  return 'import/%s:0' % node_name
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def _LoadRemoteGraphInputs(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    parent_graph_name_extended: Text,
+    remote_graph_name_extended: Text,
+    parent_graph_to_remote_graph_input_name_mapping:
+        Mapping[Text, Mapping[Text, Mapping[Text, Text]]],
+    graph_names: Set[Text]) -> beam.pvalue.PCollection:
+  """Loads remote graph inputs.
+
+  Before executing a remote graph, we need to prepare for its inputs. We first
+  get the mapping from remote graph placeholder names to parent graph input
+  names. Then, we copy the inputs from the parent graph to the remote graph
+  in element.
```

Comment:
Mention that you make a copy of the inputs, and then update the relevant key in the new copy, and then output it.

Reply:
Changed.